make-initrd: use closureInfo again

xaverdh opened the pull request:
nix 1.* is quite old now (the original issue https://github.com/NixOS/nixpkgs/issues/36268 was closed many years ago)

## Things done

This is basically a revert of 165b32d3863889f8c3615c676a201c889d860da6

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

xaverdh committed:
make-initrd: use closureInfo

philiptaron commented on pkgs/build-support/kernel/make-initrd.nix:
```diff
@@ -121,14 +119,7 @@ stdenvNoCC.mkDerivation (
-      ("closure-${toValidStoreName (baseNameOf x.symlink)}-${toString i}")
-      x.object
-    ]) contents (lib.range 0 (lib.length contents - 1));
-    pathsFromGraph = ./paths-from-graph.pl;
```
I think you can delete this file now, and remove it from `all-packages` with a throwing alias.

philiptaron reviewed (commented):
The diff looks good to me. How should it be tested? I ran a few nixos tests, since it _looks_ like this is version of making the initrd is the one used there. They all came out fine.

@mweinelt told me that things that make all nixos tests rebuild ought to go to staging, even if they have small rebuild counts here.

xaverdh committed:
pkgs/pathsFromGraph: drop

Nix 2.* is widely used now, so closureInfo should be used instead.

xaverdh commented:
> The diff looks good to me. How should it be tested? I ran a few nixos tests, since it _looks_ like this is version of making the initrd is the one used there. They all came out fine.
> 
> @mweinelt told me that things that make all nixos tests rebuild ought to go to staging, even if they have small rebuild counts here.

retargeted it at staging and removed pathsFromGraph

I tested it by building my own `initrd` both from a local checkout of nixpkgs and then cherry-picking the changes here and building from that. When comparing the outputs with `diff` they were bit-for-bit identical actually.

xaverdh commented:
Also when testing, make sure the `initrd` you are building actually uses this code path and not `make-initrd-ng`, as e.g. systems with `boot.initrd.systemd.enable = true`.

philiptaron reviewed (dismissed):
I'm sad to report that you've triggered the following bug in Nix:

- https://github.com/NixOS/nix/issues/11503

For details, read:

- https://github.com/NixOS/nixpkgs/pull/369694

To show this, run the `nixosTests.installer.simple` test. With this PR, it fails with an inscrutable `curl` error (per the bug).

How to fix? I think it's something like:

- https://github.com/NixOS/nixpkgs/pull/369151#issuecomment-2565287615

But I don't see the line that's triggering the bug immediately.

Mic92 commented:
@philiptaron how do you see that it triggers: https://github.com/NixOS/nix/issues/11503 ?

ElvishJerricco commented:
@philiptaron I *believe* the issue isn't https://github.com/NixOS/nix/issues/11503 but rather just an ordinary case of a missing dependency in the installer test. `closureInfo` depends on `jq`, and while the `installation-device.nix` module adds `jq` to `system.extraDependencies`, it's not adding the right output. It needs to add `lib.getDev jq` instead. This appears to fix it:

```diff
diff --git a/nixos/modules/profiles/installation-device.nix b/nixos/modules/profiles/installation-device.nix
index a4e5a4aac790..f01867c7e29c 100644
--- a/nixos/modules/profiles/installation-device.nix
+++ b/nixos/modules/profiles/installation-device.nix
@@ -102,7 +102,7 @@ with lib;
         stdenv
         stdenvNoCC # for runCommand
         busybox
-        jq # for closureInfo
+        (lib.getDev jq) # for closureInfo
         # For boot.initrd.systemd
         makeInitrdNGTool
       ];
```

Alternatively, we could append `jq.all` to that `extraDependencies` list just to make sure we always have the whole thing.

philiptaron commented:
> @philiptaron how do you see that it triggers: [NixOS/nix#11503](https://github.com/NixOS/nix/issues/11503) ?

Jörg, @Mic92, My evidence is circumstantial: most triggers of this bug cause `curl` to be erroneously redownloaded, and that's what I saw here when running the `nixosTests.installer.simple` test. Per Will, @ElvishJerricco, there are multiple ways to hit that, and my sniff test failed in this instance.

ElvishJerricco reviewed (dismissed):
See above comment about `jq` for the installer tests.

xaverdh committed:
nixos/installation-device: add jq.all to extraDependencies

As discovered in https://github.com/NixOS/nixpkgs/pull/372931, we need the dev output of jq for closureInfo. We opt to add the whole thing.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/issue-building-linux-kernel-modules-after-flake-update/62322/8
